### DISCUSSION TITLE
[튜토리얼] 안드로이드 휴대폰에 프로그램 설치하기

### DISCUSSION THREAD

**Discussion by airincast** [munsy0227/Chzzk-Rekoda#17] [General]
**[Read the Tutorial in English](https://github.com/munsy0227/Chzzk-Rekoda/discussions/18)**

## [필독: 설치 및 실행하기 전에 주의]
녹화한 영상은 사적인 용도 및 스트리머가 지정한 범위 외로는 **무단으로 사용할 수 없으며**, 영상 사용에 대한 **모든 책임은 녹화한 사용자에** 있습니다.

**동영상 전체를 업로드하는 행위(공개 또는 비공개에 국한되지 않음)는 저작권법 위반으로, 저작권자가 소송을 제기할 수 있습니다.**
**딥페이크 방지를 위해, AI 학습자료로 사용하는 행위 역시 금지될 수 있습니다.**

예) [라디유](https://youtube.com/channel/UC44QE3DVUuDrhUC5yp4zRYQ)의 경우
- **저작권은 라디유의 의사로 [샌드박스 네트워크](https://sandbox.co.kr/)에 위임되었습니다.**
- **AI 학습 자료로 사용하는 행위 역시 저작권 침해입니다.**
- 클립은 공식 팬 커뮤니티를 제외하고는 **다른 곳에 올릴 수 없습니다**. (공개 또는 비공개에 국한되지 않음)
- 가이드라인을 준수했다 해도 라디유 또는 샌드박스 네트워크 측의 판단으로 삭제될 수 있습니다.

**Android Q (10) 이상의 버전에서는 설치 및 실행이 어렵습니다.**
S (12) 이상의 버전에서는 별도 설정을 통해 우회할 수 있습니다. Q (10), R (11)에서는 우회가 불가능합니다.

**스트리머의 설정에 따라 대한민국 밖의 지역에서는 녹화가 차단될 수 있습니다.**

## [Termux](https://f-droid.org/ko/packages/com.termux/) 다운로드
링크된 사이트를 통해 Termux를 받아 설치합니다. (구글 플레이에서는 업데이트가 중단됐으므로 추천하지 않습니다.)

출처를 알 수 없는 앱이니만큼 설정에 따라 차단될 수도 있습니다.
이 경우 설정에서 '출처를 알 수 없는 앱 설치'를 허용한 뒤 설치 시 '무시하고 설치'를 선택합니다.

## 저장소 위치 변경하기
프롬프트에서 다음 명령어를 입력하여 패키지를 다운받을 저장소의 위치를 선택합니다.

    termux-change-repo
"Default Repositories (CF)" (CloudFlare 기반 저장소)의 사용이 권장됩니다. "Default Repositories" (기본 저장소)는 추천되지 않습니다.

## 패키지 업데이트 및 설치
패키지를 업데이트 할 때에는 apt나 pkg 중 하나를 골라 다음과 같이 입력합니다. **가급적 자주** 하는 것이 좋습니다.

    apt update && apt upgrade
    pkg update && pkg upgrade

## 핵심 패키지 설치
설치 및 실행에 필요한 모든 패키지를 설치합니다. apt 또는 pkg를 이용해 아래와 같이 입력합니다.

    apt install git uv python binutils rust libxslt libxml2 libiconv ffmpeg
    pkg install git uv python binutils rust libxslt libxml2 libiconv ffmpeg
필요 패키지에 대한 설명은 아래와 같습니다.
- git: Git 저장소에서 패키지를 가져올 때 필요
- uv: pip 대용으로 사용됨
- python: Chzzk-Rekoda의 구현 언어로, 설치, 설정, 녹화에 모두 필요 (3.10+ 필요)
- binutils, rust: Chzzk-Rekoda의 의존 패키지를 빌드하기 위한 패키지들
- libxslt, libxml2, libiconv: Chzzk-Rekoda의 의존 패키지가 따로 의존하는 패키지들
- ffmpeg: Chzzk-Rekoda에서 녹화 및 인코딩을 진행할 때 필요

설치되는 python의 버전은 2025년 1월 기준 3.12 버전입니다.

## Git 저장소 가져오기
다음 명령을 입력해 Git 저장소에서 패키지를 가져옵니다.

    git clone https://github.com/munsy0227/Chzzk-Rekoda
이후에는 아래와 같이 입력해 디렉터리 이름을 변경합니다. 필수는 아니나 **가급적 소문자로** 할 것을 권장합니다.

    mv Chzzk-Rekoda chzzk
최신 버전으로 업데이트 할 때에는 디렉토리에 들어가 아래 명령들을 순차적으로 입력합니다.

    git fetch && git stash && git rebase && git stash pop
**[의존 패키지 설치, 채널 설정, 녹화 실행은 README.md 파일에 기술](https://github.com/munsy0227/Chzzk-Rekoda/blob/main/README.md)되어 있습니다.**

## 영상 파일 접근
termux에서 스토리지 접근을 허용하기 위해서는 아래와 같이 입력합니다.

    termux-setup-storage
이후에는 파일 접근을 위해 주어진 2개의 옵션을 사용합니다.

    am start -a android.intent.action.VIEW -d "content://com.android.externalstorage.documents/root/primary"
추가 조작이 필요하지만 파일 이동/복사/삭제가 지원됩니다.

    am start -a android.intent.action.OPEN_DOCUMENT -d /storage/emulated/0 -t '*/*'
파일을 바로 플레이해 볼 수 있지만 파일 이동/복사/삭제는 지원하지 않습니다.

**Comment by airincast** [comment]
## [문제 해결: 녹화하다가 중간에 뻗어버렸다면? 이를 방지하려면?]
이 문서에서 서술하는 팁은 Android S(12) 및 그 이상의 버전에서 적용할 수 있습니다.
**Android Q(10), R(11)에선 적용이 불가능**하며, P(9) 이하인 경우에는 적용하지 않아도 됩니다.

이 문서에서는 사전에 Android Studio 또는 Command Tools의 설치를 마쳤음을 가정합니다.

안드로이드에서는 **자체적으로 Phantom Process의 수를 제한**하고 있으며
이 개수를 넘어가면 안드로이드가 임의로 프로세스를 종료합니다.
이를 해결하려면 임시로 **프로세스 개수 제한을 풀어야** 합니다.

휴대폰에서 **설정**에서 **개발자 모드**로 들어가 **USB 디버깅**을 켭니다.

이후에는 휴대폰을 USB로 PC에 연결합니다. **디버깅 허용여부가 나타나면 허용**합니다.

<SDK 디렉토리>/platform-tools 디렉토리에 들어가 아래의 명령어를 순차적으로 입력합니다.

    adb shell "/system/bin/device_config set_sync_disabled_for_tests persistent"
    
device_config 서비스의 플래그 동기화를 일시적으로 방지합니다. 복구 시에는 **persistent** 대신 **none** 을 사용합니다.

    adb shell "/system/bin/device_config put activity_manager max_phantom_processes 2147483647"
허용하는 Phantom Process의 개수를 최대치(약 21억개)로 올립니다. 기본 값은 **32** 입니다.

    adb shell settings put global settings_enable_monitor_phantom_procs false
Phantom Process에 대한 모니터링을 끕니다. 기본 값은 **false** 가 아니라 **true** 입니다. (모니터링 사용)

설정이 끝났다 해도 다른 프로세스의 영향을 받을 수 있습니다.  
녹화 전에는 가급적 **백그라운드 프로세스를 모두 종료**해 주세요.

**Comment by apppeter1** [comment]
어제 잘못 녹화된 파일이 있어서 am start -a android.intent.action.VIEW -d "content://com.android.externalstorage.documents/root/primary" 명령어를 사용했는데 연란처가 열리는게 정상인가요? 밑에 명령어로는 파일이 안지워지더라구요
